test_runner/src/main/kotlin/ftl/reports/util/ReportManager.kt

```diff
@@ -81,6 +82,9 @@ object ReportManager {
 
         JUnitReport.run(matrices, testSuite)
 
+        val localJunitXmlPath = JUnitReport.reportPath(matrices)
+        GcStorage.uploadJunitXml(localJunitXmlPath, args)
```

Comment:
I think this should only be uploaded if all tests passed.

Reply:
In the algorithm part of this feature, we'll be ignoring failed tests. Does that address your feedback?

It's exceptionally rare for all tests to pass on large test suites. If 99.99% of tests pass and one fails, then there's a lot of value in using the timing data from the passed tests. I'm not sure  discarding all test results if a single failure is present makes sense.

Reply:
I think it's perfectly fine to toss the whole thing if tests are failing. Think of it this way. There's a `master` branch where _ALL_ tests are passing, and to merge into `master`, you need to pass tests anyway.

Failing tests just means you don't go update a perfectly working dataset of test times with your failing test data.

Fix your tests and it'll go update the timing xml.

Reply:
The only problem I see is there are test suites where the tests never pass 100%. At that point the smart flank feature does nothing for them.

Let's have an example scenario of 1,000 tests with 1 failure. The failure is different each time because UI tests are flaky. In the 0 failures allowed model, smart flank does nothing until the test suite passes completely. In the always upload model, the test suite benefits regardless of if the suite perfectly executed.

I'm not sure why we wouldn't always want the benefit, even if sometimes it'll be partially beneficial. I agree with disregarding individual failing tests.

Reply:
It just needs to pass once. For an entire repository. As the location of the file is not variable between branches/builds, we need to be safe.

Imagine a scenario where _all_ tests fail. And it just happily goes and updates the xml file with wonky test times, maybe zeros.

Now, not only are the tests for your branch messed up. You also messed up sharding for the rest of your team.

Reply:
Imho, It's not flank's responsibility to worry about cases where a team has flaky tests.

Smart flank will work if you have ONE successful test run.

Reply:
I don't think that's how [Smart Flank works in the Java code](https://github.com/TestArmada/flank/tree/flank_java). I'm worried about users getting surprised by this behavior after upgrading.

With the CLI option coming to flank, it's easy to have multiple timing files. You could dynamically define the timing by branch:

- `flank android run --junitGcsPath gs://myapp/flank/$BRANCH_NAME.xml`

> Now, not only are the tests for your branch messed up. You also messed up sharding for the rest of your team.

Ok, let's say a bad branch was pushed with all failing tests. Sharding is messed up. As soon as the next run happens, the timing info is good again.

Now let's say we have a flaky test suite of ~99% with no 100% successful runs. The timing info will never be good. There are a lot of companies out there with UI test suites that don't pass. Unfortunately flakiness is the norm.

Flank should do the right thing for the majority of users by default. While also [providing an API in the future](https://github.com/TestArmada/flank/issues/286) for users to customize behavior. I could see, for example, a team that has test suites that pass 100% wanting to enforce that for JUnit XML uploads. I'm not persuaded that makes sense for the default option.

Reply:
I think if we had a `--no-upload` option that might be useful. Local runs/WIP PRs could then optionally use the existing XML without having to always upload a new one.

Reply:
I find it hard to believe an organization would not devote some time to tests which _always_ fail ( which is the only scenario where you would want to upload without checking for success. )

The thing is, when you upload without ensuring success, the timing info is _never_ good. You always shard off of potentially bad data. If a test is failing, it might fail in 0.5 seconds. This would only make the implementation wonky as it raises questions like - 

- If a test has failed fast in our reference junit xml, do I shard it using its very low time value? Even though it might be a junk value as it failed quickly? For example a test may fail in 0.5s when it usually takes 3s. Now you're sharding subsequent tests based off of wrong data anyway.

- If the test hangs, you might have large values. How do you handle that?

- Are you just going to ignore failed tests when it comes to deciding shard assignment?

Reply:
Alternatively, just make storing regardless of failure a configurable option. gRPC is coming. But that's sometime in Q1/Q2 2019.

Reply:
> I don't think that's how Smart Flank works in the Java code. I'm worried about users getting surprised by this behavior after upgrading.

Not necessarily a bad thing. Flank java was never very good with this as Firebase's junit was messed up. So everytime a new test was added, they'd single shard the whole test suite.

I remember trying to do that, and it took forever. So I scrapped time based sharding alltogether. And We have one of the smaller test suites out there.

Reply:
> In the algorithm part of this feature, we'll be ignoring failed tests. Does that address your feedback?

I only just saw this. 🤦‍♂️ 

In a way, yes. I suggest we do some extra work and only replace/add the test times for non-failing tests. So instead of uploading the firebase provided junit xml, we could process and upload our own or even use json.

This would be the best of both worlds. As if a test is failing, its time is not updated. But a time from when it _did_ work isn't lost either.

Reply:
> The thing is, when you upload without ensuring success, the timing info is _never_ good. You always shard off of potentially bad data.

I think the timing info for tests that passed is good? The XML includes passed/failed/skipped. If 1 out of 1000 fails then I have good timing data for 999 tests. The failure is discarded.

> * Are you just going to ignore failed tests when it comes to deciding shard assignment?

I think so. This is an existing feature in Flank Java that worked well at scale. No other test runner that I'm aware of requires perfect pass rates to take advantage of test timing information. We have to ignore new tests anyway since they won't have previous timing data.

If we discover this is a real problem I'm happy to revisit in the future. With Flank Java, we didn't have any such reports.

Reply:
> If we discover this is a real problem I'm happy to revisit in the future. With Flank Java, we didn't have any such reports.

Probably because it wasn't used much. 🙈

Reply:
Here are the options as I understand them:

- Upload only on 100% pass rate
    - Pros
        - Timing info is accurate for all tests on a perfect run
    - Cons
        - Timing info is out of date if the last run wasn't a perfect run
        - Unhealthy test suites (< 100% pass rate) don't benefit from time data

- Always upload
    - Pros
        - Timing info is accurate for passed tests
    - Cons
        - Not all tests may have passed.
        - Test failures remove test timing info from the XML

- Always upload + merge success
    - Pros
        - Timing info is accurate for passed tests.
        - If a test failed previously, then we keep the last known good timing
    - Cons
        - Additional complexity to implement

Reply:
Option 3 has complexity at the implementation ( and not _that_ much either ). Option 1 and 2 are detrimental to the workflows of the consumers. A subset in each case. ( 2 in my case. 1 in yours )

2 especially so (imo) , as _some_ tests may fail, or _all_ tests may fail. Making sharding properly the exception rather than the norm due to the inconsistency of the data.

Reply:
I can help with the implementation of 3.

Reply:
I agree that `3) Always upload + merge success`  makes the most sense.